lib/diskoverflow/common.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Package diskoverflow provides several data container types which are limited
+// in their memory usage. Once the total memory limit is reached, all new data
+// is written to disk.
+// Do not use any instances of these types concurrently!
+package diskoverflow
+
+import (
+	"sync"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+const OrigDefaultOverflowBytes int64 = 16 << protocol.MiB
+
+var (
+	defaultOverflow    int64 = OrigDefaultOverflowBytes
+	defaultOverflowMut       = sync.Mutex{}
+)
+
+// Sets the default limit of when data is written to disks. Can be overruled
+// on individual instances via their SetOverflowBytes method.
+// Argument bytes < 0 means no disk spilling will ever happen and = 0 sets it to
+// OrigDefaultOverflowbytes.
+func SetDefaultOverflowBytes(bytes int64) {
+	defaultOverflowMut.Lock()
+	if bytes == 0 {
+		defaultOverflow = OrigDefaultOverflowBytes
+	} else {
+		defaultOverflow = bytes
+	}
+	defaultOverflowMut.Unlock()
+}
+
+func DefaultOverflowBytes() int64 {
+	defaultOverflowMut.Lock()
+	defer defaultOverflowMut.Unlock()
+	return defaultOverflow
+}
+
+// Value must be implemented by every type that is to be stored in a disk spilling container.
+type Value interface {
+	Size() int64
+	Marshal() []byte
+	Unmarshal([]byte) Value // The returned Value must not be a reference to the receiver.
```

Comment:
I changed the `Unmarshal` to modify itself instead of returning it. The error I wanted to avoid, because everywhere I remember we just panic if there is an error marshalling. And I also wouldn't know how to treat a marshal failure any other way than by panicking.